### PR TITLE
Run fusion property test with arbitrary layouters

### PR DESCRIPTION
**Pull request by sjakobi**
Context: #95

### CODE REVIEW

File: prettyprinter/test/Testsuite/Main.hs [hunk 1]
```diff
@@ -86,20 +89,59 @@ tests = testGroup "Tests"
...
+    [ pure layoutPretty
+    , pure layoutSmart
+    , pure (const layoutCompact)
+    -- , Internal.layoutWadlerLeijen <$> arbitrary -- too inconsistent for fusionDoesNotChangeRendering
+    ]
+
+instance Arbitrary LayoutOptions where
+    arbitrary = LayoutOptions <$> oneof
+        [ AvailablePerLine <$> arbitrary <*> arbitrary
+        -- , pure Unbounded -- https://github.com/quchen/prettyprinter/issues/91
```
Comment (sjakobi) [inline prettyprinter/test/Testsuite/Main.hs]: If I uncomment this, the fusion tests trigger #91:

```
    Shallow fusion does not change rendering: FAIL (0.02s)
      *** Failed! (after 56 tests):
      Exception:
        »SFail« must not appear in a rendered »SimpleDocStream«. This is a bug in the layout algorithm! Please report this as a bug
        CallStack (from HasCallStack):
          error, called at src/Data/Text/Prettyprint/Doc/Render/Util/Panic.hs:13:21 in prettyprinter-1.5.1-inplace:Data.Text.Prettyprint.Doc.Render.Util.Panic
      oakland <(commence   inverse)>buttonaccrue tapewormtrackercrowfoot
      spheroid ( printer ` ""sailboatbackfieldendorse transit gazelletacticsdrumbeat  dogsled  dropper puppy dashboard  
       skydive miser eyeglass apple guidancetonic
      oakland southward )
      woodlark
      minnow
      slowdown
      <function>
      LayoutOptions {layoutPageWidth = Unbounded}
      Exception thrown while showing test case:
        »SFail« must not appear in a rendered »SimpleDocStream«. This is a bug in the layout algorithm! Please report this as a bug
        CallStack (from HasCallStack):
          error, called at src/Data/Text/Prettyprint/Doc/Render/Util/Panic.hs:13:21 in prettyprinter-1.5.1-inplace:Data.Text.Prettyprint.Doc.Render.Util.Panic
      
      Use --quickcheck-replay=857723 to reproduce.
```

Two things could IMHO help analyzing these failures:

1. Shrinking: I wonder whether it might be easiest to just use `hedgehog` instead of `QuickCheck`, since that has built-in shrinking. What do you think about this @quchen?

2. A diagnostic representation of the `Doc` (#90).

**Comment by quchen** [comment]
I haven’t used Hedgehog and I’m fairly familiar with Quickcheck, but certainly not opposed to using new technology in order to improve tests. Automatic shrinking does sound cool, I should read up on that!

File: prettyprinter/test/Testsuite/Main.hs (hunk 1 above)
Comment (quchen) [reply prettyprinter/test/Testsuite/Main.hs]: The `arabitrary` should be explicit `range`s, otherwise you get ribbons of 1e-5 and 2 for example.

Comment (sjakobi) [reply prettyprinter/test/Testsuite/Main.hs]: > The `arabitrary` should be explicit `range`s, otherwise you get ribbons of 1e-5 and 2 for example.

Why would that be a problem? I don't really understand the ribbon width logic so far, but I know that `layoutWadlerLeijen` clamps the result to a valid range:

https://github.com/quchen/prettyprinter/blob/7da3b1d32f8a74efb3bcf1b2d062f8f24a64c918/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L1810-L1814

I usually try to not to constrain the generated values unless they cause problems – "nonsense" inputs are often helpful at finding bugs in edge cases.

**Comment by sjakobi** [comment]
I'd like to build on this in my PRs, so I'll merge soon. Further improvements will hopefully happen as I use these tests.

Regarding shrinking: `genericShrink` will probably break invariants and lead to the wrong problems. Maybe I should write a shrinker by hand. `hedgehog` still seems attractive too…
